src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -216,6 +216,53 @@ _mongoc_topology_scanner_cb (uint32_t id,
    bson_mutex_unlock (&topology->mutex);
 }
 
+static void
+_server_session_init (mongoc_server_session_t *session,
+                      mongoc_topology_t *unused,
+                      bson_error_t *error)
+{
+   _mongoc_server_session_init (session, error);
+}
+
+static void
+_server_session_destroy (mongoc_server_session_t *session,
+                         mongoc_topology_t *unused)
+{
+   _mongoc_server_session_destroy (session);
+}
+
+static int
+_server_session_should_prune (mongoc_server_session_t *session,
```

Comment:
If I read correctly, this is only being called on a particular session while retrieving it, right? If so, is there a risk of expired sessions hanging around indefinitely if they just happen to not get touched? If so, should there be a backstop to occasionally still go through and remove the expired threads?

Super cool if there's no sane place or time to do so; I know the point is to reduce contention on the below mutex anyway so an even occasional search-through might not work.

Reply:
This was discussed, yes. And indeed, a expired sessions can loom indefinitely in a pathological case. We also figure that the pool itself can never grow indefinitely, since creating a new element requires that all stale elements be deleted first. It *may* be the case that:

1. A burst of 500 simultaneous operations generate 500 session objects.
2. Those sessions are returned to the pool.
3. For an indefinite amount of time, only one session is used at a time, and the other 499 sit unused.
4. The bottom 499 sessions expire, but are not deallocated, as the top session is the only one ever inspected.
5. If at any attempt is made to use *two* sessions simultaneously, it will immediately prune the 499 sessions and the pool will be cleared.

This may have tradeoffs, but because of the LIFO order, the moment a single expired session is popped from the top, the entire pool will be cleared. We discussed in another comment on how to reconcile this with the spec that mandates the "bottom" of the pool be checked every time a session is returned. If we want to go this route, the pool will need to be tweaked into a deque, and that's very possible in the future as an upcoming change will require a thread-safe deque anyway.